instrumentation/apache-httpasyncclient-4.1/src/main/java/io/opentelemetry/javaagent/instrumentation/hypertrace/apachehttpasyncclient/ApacheAsyncClientInstrumentationModule.java

```diff
@@ -95,70 +103,113 @@ public int order() {
   }
 
   public static class HttpAsyncClient_execute_Advice {
+
     @Advice.OnMethodEnter(suppress = Throwable.class)
     public static void enter(
         @Advice.Argument(value = 0, readOnly = false) HttpAsyncRequestProducer requestProducer,
         @Advice.Argument(value = 2) HttpContext httpContext,
-        @Advice.Argument(value = 3, readOnly = false) FutureCallback futureCallback) {
-      if (requestProducer instanceof DelegatingRequestProducer) {
-        DelegatingRequestProducer delegatingRequestProducer =
-            (DelegatingRequestProducer) requestProducer;
-        Context context = delegatingRequestProducer.getContext();
-        requestProducer = new DelegatingCaptureBodyRequestProducer(context, requestProducer);
-        futureCallback = new BodyCaptureDelegatingCallback(context, httpContext, futureCallback);
+        @Advice.Argument(value = 3, readOnly = false) FutureCallback<?> futureCallback) {
+      if (requestProducer instanceof DelegatingRequestProducer
+          && futureCallback instanceof WrappedFutureCallback<?>) {
+        Context context = currentContext();
+        WrappedFutureCallback<?> wrappedFutureCallback = (WrappedFutureCallback<?>) futureCallback;
+        BodyCaptureDelegatingCallback<?> bodyCaptureDelegatingCallback =
+            new BodyCaptureDelegatingCallback<>(context, httpContext, futureCallback);
+        futureCallback = bodyCaptureDelegatingCallback;
+        requestProducer =
+            new DelegatingCaptureBodyRequestProducer(
+                context, requestProducer, wrappedFutureCallback, bodyCaptureDelegatingCallback);
       }
     }
   }
 
   public static class DelegatingCaptureBodyRequestProducer extends DelegatingRequestProducer {
 
-    final Context context;
+    private static final Logger log =
+        LoggerFactory.getLogger(DelegatingCaptureBodyRequestProducer.class);
+
+    private static final Lookup lookup = MethodHandles.lookup();
+    private static final MethodHandle getContext;
+
+    static {
+      MethodHandle localGetContext;
+      try {
+        Field contextField = WrappedFutureCallback.class.getDeclaredField("context");
```

Comment:
Could you please create an issue in the OTEL upstream to expose the context and add TODO comment here with the link.

Reply:
Unfortunately, I'm not sure that it's a reasonable expectation for them to do this. Before, we accessed the context via a public access to a `final` field on `DelegatingRequestProducer`. However, in [OTEL java-instrumentation pull 3209](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/3209/files) a change was made to postponing http client span creation until the request is available. This meant that the `DelegatingRequestProducer` no longer has a client context when it is instantiated, but rather sets a new `volatile` field on `WrappedFutureCallback`. I don't think it's a reasonable ask to expose this volatile field, and I can see why they would not want to expose a nullable volatile field to downstream instrumentation libraries as it can only be invoked after the `DelegatingRequestProducer.generateRequest`. IMO I think we should refactor our implementation of this instrumentation to not rely on this field and instead be able to do something like `Java8BytecodeBridge.currentSpan()` but for some reason I can't get that to work properly. I'll add a comment indicating this